File: New folder/cdp-sdk-8fcfac348487f586b224b77074f52594ad44198b/examples/python/evm/spend-permissions/revoke_spend_permission.py
# Usage: uv run python evm/spend-permissions/revoke_spend_permission.py

import asyncio

from cdp import CdpClient
from cdp.spend_permissions import SpendPermissionInput
from cdp.utils import parse_units

from dotenv import load_dotenv

load_dotenv()


async def main():
    """Main function to demonstrate creating and revoking a spend permission."""

    async with CdpClient() as cdp:
        account = await cdp.evm.get_or_create_smart_account(
            name="Example-Account-Revoke",
            owner=await cdp.evm.get_or_create_account(
                name="Example-Account-Revoke-Owner",
            ),
            enable_spend_permissions=True,
        )
        print(f"Account Address: {account.address}")

        # Create a spender account
        spender = await cdp.evm.create_account()
        print(f"Spender Address: {spender.address}")

        spend_permission = SpendPermissionInput(
            account=account.address,
            spender=spender.address,
            token="usdc",
            allowance=parse_units("0.01", 6),
            period_in_days=1,
        )

        # Create the spend permission onchain
        user_operation = await cdp.evm.create_spend_permission(
            spend_permission=spend_permission,
            network="base-sepolia",
        )
        print(
            f"Created spend permission with user operation hash: {user_operation.user_op_hash}"
        )

        # Wait for the user operation to complete
        result = await cdp.evm.wait_for_user_operation(
            smart_account_address=account.address,
            user_op_hash=user_operation.user_op_hash,
        )
        print(f"User operation completed with status: {result.status}")

        # List the spend permissions
        permissions = await cdp.evm.list_spend_permissions(account.address)

        # Revoke the spend permission
        revoke_user_operation = await cdp.evm.revoke_spend_permission(
            address=account.address,
            permission_hash=permissions.spend_permissions[0].permission_hash,
            network="base-sepolia",
        )

        # Wait for the revoke user operation to complete
        revoke_result = await cdp.evm.wait_for_user_operation(
            smart_account_address=account.address,
            user_op_hash=revoke_user_operation.user_op_hash,
        )
        print(f"Revoke user operation completed with status: {revoke_result.status}")


if __name__ == "__main__":
    asyncio.run(main())

File: New folder/cdp-sdk-8fcfac348487f586b224b77074f52594ad44198b/examples/typescript/evm/spend-permissions/revokeSpendPermission.ts
// Usage: pnpm tsx evm/spend-permissions/revokeSpendPermission.ts
import {
  CdpClient,
  parseUnits,
  type SpendPermissionInput,
} from "@coinbase/cdp-sdk";
import "dotenv/config";
import { Hex } from "viem";

const cdp = new CdpClient();

const account = await cdp.evm.getOrCreateSmartAccount({
  name: "Example-Account-Revoke",
  owner: await cdp.evm.getOrCreateAccount({
    name: "Example-Account-Revoke-Owner",
  }),
  enableSpendPermissions: true,
});

const spender = await cdp.evm.createAccount();

const spendPermission: SpendPermissionInput = {
  account: account.address,
  spender: spender.address,
  token: "usdc",
  allowance: parseUnits("0.01", 6),
  periodInDays: 30, // monthly
};

const { userOpHash } = await cdp.evm.createSpendPermission({
  spendPermission,
  network: "base-sepolia",
});

const userOperationResult = await cdp.evm.waitForUserOperation({
  smartAccountAddress: account.address,
  userOpHash,
});

console.log("User Operation:", userOperationResult);

const permissions = await cdp.evm.listSpendPermissions({
  address: account.address,
});

const permissionHash = permissions.spendPermissions[0].permissionHash;

const { userOpHash: revokeUserOpHash } = await cdp.evm.revokeSpendPermission({
  address: account.address,
  permissionHash: permissionHash as Hex,
  network: "base-sepolia",
});

const revokeUserOperationResult = await cdp.evm.waitForUserOperation({
  smartAccountAddress: account.address,
  userOpHash: revokeUserOpHash,
});

console.log("Revoke User Operation:", revokeUserOperationResult);
